### PR TITLE
Composer 2 + AMP Plugin 2.1 / AMP PHP Toolbox

### LINKED ISSUE #4897: Update AMP PHP Library Integration
## Feature Description

Tracking ticket for once we need to update our usage of the AMP PHP library.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #5083: Composer 2 Support
## Task Description

This is currently blocked as some dependencies need to be updated first.

https://github.com/civicrm/composer-downloads-plugin/pull/1
https://github.com/ampproject/amp-wp/issues/5551

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

* Moves the project to use Composer v2.
* ...which is only supported by AMP plugin v2.1 (currently in development)
* ...which in turn requires using AMP PHP toolbox from Packagist

## Relevant Technical Choices

Note: This supersedes #5284

Since AMP plugin 2.1 hasn't been released yet, this loads the plugin from `dev-develop` instead.

## To-do

* Testing
* Reduce dependency on AMP plugin itself so we can rely solely on AMP PHP Toolbox

## User-facing changes

N/A

## Testing Instructions

N/A

---

Fixes #5083 
Fixes #4897

### CODE REVIEW

**Comment by spacedmonkey** [comment]
This didnt update me to composer 2. 

```
You are already using composer version 1.10.17 (1.x channel).
 done!
```

**Comment by spacedmonkey** [comment]
I needed to run 
```
 composer self-update --2
```

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
I was running 1.10.17 and didnt upgrade.

**Comment by swissspidy** [comment]
> This didnt update me to composer 2.

What do you mean with _this_?

The PR just updates the plugin to be Composer 2 compatible.

**Comment by pierlon** [comment]
Update on this: the AMP plugin will be releasing v2.0.7 today, which is a hotfix for https://github.com/ampproject/amp-wp/issues/5609. Due to an unnoticed incompatibility with Composer v2, we're downgrading back to v1 until we have a solution in place to fix the reported issue.

**Comment by spacedmonkey** [comment]
@pierlon What is the status of Composer 2 in 2.0.8?

**Comment by swissspidy** [comment]
> @pierlon What is the status of Composer 2 in 2.0.8?

@spacedmonkey Still using Composer 1 on the 2.0 branch. Upgrading to Composer v2 will be revisited for 2.1+.

File: composer.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
     "ext-json": "*",
     "ext-libxml": "*",
     "ext-mbstring": "*",
-    "ampproject/amp-wp": "^2.0.10",
+    "ampproject/amp-toolbox": "^0.1",
+    "ampproject/amp-wp": "dev-develop",
```
Comment (spacedmonkey) [inline composer.json:15]: Wouldn't it be better to do `@stable` ?

Comment (swissspidy) [reply composer.json:15]: Wow you're fast!

See the PR description on why this is necessary at the moment.

In short: only AMP plugin 2.1+ will support Composer 2, but it is still in development (= no stable version to use yet)

Comment (spacedmonkey) [reply composer.json:15]: 🤦‍♂️  I didn't read it before reviewing. I personally really hate not define versions like this. I understand why it is done, but I would like it changed once we can make as stable.

**Comment by swissspidy** [comment]
@westonruter @pierlon Is there any rough ETA on the AMP plugin v2.1 release?

We're considering just using `dev-develop` and AMP Toollbox `dev-main` for the time being (see changes in this PR) so that we're not really blocked on you folks on improving our code base.

**Comment by westonruter** [comment]
@swissspidy We are planning to release v2.1 to coincide with WordPress 5.7. So roughly a month from now. If the prerelease state works for you, then I don't see any problem with that. Note that AMP Toolbox PHP should have a new stable release prior to the AMP plugin's v2.1. So you may only need to use `dev-develop` for the AMP plugin.

@pierlon Is there any dependency on https://github.com/ampproject/amp-wp/pull/5745?

**Comment by pierlon** [comment]
> @pierlon Is there any dependency on ampproject/amp-wp#5745?

Yes, it would be of high priority now. At the moment if our patched version of the `sabberworm/php-css-parser` is not autoloaded first, it will impact the how the AMP plugin works internally. This wouldn't impact the Web Stories plugin, however, as it currently bundles the patched version of the CSS parser library and is scoped accordingly so that it always uses the one its shipped with.

File: docs/getting-started.md [hunk 2]
```diff
@@ -4,8 +4,8 @@
 
 To contribute to this plugin, you need the following tools installed on your computer:
 
-- [PHP](https://www.php.net/) - version 7.2 or higher, preferably installed via [Homebrew](https://brew.sh/)
-- [Composer](https://getcomposer.org/) (PHP package manager) - to install PHP dependencies. Preferably version 1.x (see #5083 for Composer v2 support).
+- [PHP](https://www.php.net/) - version 7.2 or higher, preferably installed via [Homebrew](https://brew.sh/).
```
Comment (spacedmonkey) [inline docs/getting-started.md:7]: In my testing, it didn't work in PHP 8.0. We should document this here. 

Screenshot 2021-02-18 at 10 52 08

Comment (swissspidy) [reply docs/getting-started.md:7]: That's unrelated to this PR?

PHP-Scoper will fail on PHP 8 also on `main`

**Review by spacedmonkey — APPROVED** [review]
Worked well in PHP 7.3 and PHP 7.4 I tested locally. However, got error when installing on PHP 8.0. This is an issue in main, so I am going to make another issue to fix this elsewhere.

File: docs/getting-started.md (hunk 2 above)
Comment (swissspidy) [reply docs/getting-started.md:7]: We should just update PHP-Scoper to 0.14.0, which has PHP support.

**Comment by swissspidy** [comment]
> Worked well in PHP 7.3 and PHP 7.4 I tested locally. However, got error when installing on PHP 8.0. This is an issue in main, so I am going to make another issue to fix this elsewhere.

I just updated PHP-Scoper here which should fix this. Could you try again?

**Review by spacedmonkey — APPROVED** [review]
Works in PHP 8.0.

**Pull request merged**